### PR TITLE
Corrections for the MVP

### LINKED ISSUE #5: function can be a method of Registers struct
https://github.com/iinaki/lc-3-vm/blob/6bd3cc03680acfaa228826c1fce7572eb71e1e6d/src/utils.rs#L41

### LINKED ISSUE #6: All opcodes should be methods of the VM struct instead of separate functions.

### PR DESCRIPTION
**Pull request by iinaki**
## Changes
- Changed the way errors where handled, implemented a `VmError` struct for this.
- Removed code from `mod.rs`.
- Moved `update_flags` function to be a method of `Vm`.
-  Changed  every opcode to be a methods of `Vm` instead of separate functions. For this I implement each opcode in its own module separately as a method of the `Vm`.

### CODE REVIEW

**Comment by JereSalo** [comment]
I probably wouldn’t have defined traits for every opcode. Can’t they just be each on their own file without the need for them to be traits? What do you think would be the disadvantage of just doing `impl VM` on each opcode file without creating traits? I’m just curious about that. I think it would still look clean without unnecessary traits. What are your thoughts?

File: src/registers.rs [hunk 1]
```diff
@@ -104,9 +108,28 @@ impl Registers {
...
+    ///
+    pub fn update_flags(&mut self, r: u16) -> Result<(), VmError> {
+        if self.get(r)? == 0 {
+            self.cond = FL_ZRO;
+        } else if (self.get(r)? >> 15) & 1 == 1 {
+            self.cond = FL_NEG;
+        } else {
+            self.cond = FL_POS;
+        }
+        Ok(())
```
Comment (JereSalo) [inline src/registers.rs]:
```suggestion
    pub fn update_flags(&mut self, r: u16) -> Result<(), VmError> {
        let r_value = self.get(r)?;
        if r_value == 0 {
            self.cond = FL_ZRO;
        } else if (r_value >> 15) & 1 == 1 {
            self.cond = FL_NEG;
        } else {
            self.cond = FL_POS;
        }
        Ok(())
```

I would change update_flags to look something like this. It is not important though. I'm not even sure if it looks better this way 😛

**Review by JereSalo — APPROVED** [review]
I didn’t see anything alarming; the corrections are well implemented overall. The suggestions that I made aren't crucial, just things that caught my attention. There's no need to make changes, I just wanted to know your point of view.

**Comment by iinaki** [comment]
> I probably wouldn’t have defined traits for every opcode. Can’t they just be each on their own file without the need for them to be traits? What do you think would be the disadvantage of just doing `impl VM` on each opcode file without creating traits? I’m just curious about that. I think it would still look clean without unnecessary traits. What are your thoughts?

Done! You are right they were unnecessary, I changed to do `impl Vm` en each opcode module :)

File: src/registers.rs (hunk 1 above)
Comment (iinaki) [reply src/registers.rs]: Done! It's better this way as it uses `get` just once :)

**Review by JereSalo — APPROVED** [review]
Well done! Looks very clean. Congrats 🎉

**Pull request merged**